Firestore/core/src/core/composite_filter.cc

```diff
@@ -154,14 +154,14 @@ const model::FieldPath* CompositeFilter::Rep::GetFirstInequalityField() const {
 
 const std::vector<FieldFilter>& CompositeFilter::Rep::GetFlattenedFilters()
     const {
-  if (Filter::Rep::memoized_flattened_filters_.empty() && !filters().empty()) {
-    for (const auto& filter : filters()) {
+  return memoized_flattened_filters_->memoize([&]() {
```

Comment:
While you're doing this "migration" to thread-safe memoization, could you look into how much more effort it would be to do the other memoized values? This would (a) make the code more robust and (b) make it more consistent.

Other "memoized" member variables I found were:
- memoized_aggregate_target
- memoized_indexes_
- memoized_max_index_id_
- memoized_max_sequence_number_
- memoized_target

(also, `memoized_aggregate_target`, and `memoized_target` should probably have a trailing underscore appended to their names, if you're in there already)

Reply:
memoized_max_sequence_number_, memoized_max_index_id_ gets contantly read, compare and update. for example: 
```
int32_t memoized_max_index_id_ = -1;
int64_t memoized_max_sequence_number_ = -1;

memoized_max_index_id_ = std::max(memoized_max_index_id_, index_id);
memoized_max_sequence_number_ = std::max(memoized_max_sequence_number_, sequence_number);
```
They are more of a counter/recorder than a cached value that gets computed once and for all. 

Correct me if I am wrong, but I feel like `mutex lock` is more appropriate for them than `call_once` @dconeybe

Reply:
Actually, you can skip `memoized_indexes_`, `memoized_max_index_id_`, and `memoized_max_sequence_number_`. They are part of `LevelDbIndexManager` which is only ever used from one thread (the AsyncQueue's thread). So no synchronization is necessary.

Sorry for wasting your time on this. I should have looked more closely.

These variables aren't really named correctly IMO. The term "memoize" isn't the right term for values that are regularly updated (naming is hard).

Reply:
not at all. I got a better understanding of call_once ❤️